### ISSUE TITLE
Task Fail: Get meeting recordings from Zoom: null value in column "recordingtype" violates not-null constraint

### ISSUE THREAD

**Issue by smangancap** [jrchamp/moodle-mod_zoom#619]
Moodle 4.1.12 (Build: 20240812)
Zoom Meeting v5.2.32024072500

Get meeting recordings from Zoom (mod_zoom\task\get_meeting_recordings) fails with:

**Error: null value in column "recordingtype" violates not-null constraint** 

This is halting all other meeting recording retrievals.

Since the last update had to do with a modification to recordingtype I am wondering if there was a regression. I haven't been able to find any other bug reports or discussions around this specific error yet.

**Update**: I was able to locate the Zoom meeting activity in Moodle and I can't find anything out of the norm in the settings. It is set up to be a recurring meeting. There have been 4 sessions so far with this particular ID and the previous recordingtype values were set to chat_file. 

Also, a breakout room was created but no participants added. I tried to delete the breakout room just for the fun of it but could not. No error, the room just wouldn't delete after I saved the activity. I tried same in a different activity. Checked system logs, no nginx errors. Unsure if any of this could be related to the original problem of the task fail, but thinking more info the better.

Hoping that additional information helps.

**Comment by jrchamp** [comment]
I'm not sure what you would ever get a null value for recording type. Here's the critical line of code:
https://github.com/ncstate-delta/moodle-mod_zoom/blob/6598de0f6258b1261759a732013a10d5baf58e97/classes/webservice.php#L1073

Zoom's [Get meeting recordings](https://developers.zoom.us/docs/api/rest/reference/zoom-api/methods/#operation/recordingGet) API suggests that the `recording_type` value will always be a string from the list they provide. If you are able to debug this somewhere safely, `if (empty($recording->recording_type)) { var_dump($recording); }`

As an alternative, the crtiical line of code could be modified to default to a non-null value `$recordinginfo->recordingtype = $recording->recording_type ?? 'null';`. This will set the value to the string "null" so that the data can be loaded even if it does not have a value.

If you get more information about the problematic data element, we would be very interested to know if this is a valid data element or if there is an issue in Zoom's API.

**Comment by smangancap** [comment]
Hello, thank you so much for getting back. The problem seems to have auto-magically resolved itself this week and I can't explain what change could have occurred to resolve it. There was no back-end DB updates or configuration changes made.

**Comment by smangancap** [comment]
I spoke too soon. The errors stopped and the job ran consecutively for a number of days and in checking our sys logs again, we appear to have a new one. Same error, different recording.

**Comment by jrchamp** [comment]
Hi @smangancap, thanks for your quick response. We have a potential fix: https://github.com/ncstate-delta/moodle-mod_zoom/pull/620/commits/03c7c3b5b072f6baee2d0bb15d676157963e1c01. If the error is still occurring and you can add the code that will show the structure of the recording object, it may help explain if Zoom's API is behaving differently these days or perhaps has a bug.

```php
if (empty($recording->recording_type)) {
    var_dump($recording);
}
```

**Comment by smangancap** [comment]
I can see how this would be helpful for you to debug, but unfortunately, since we are hosted I don't have permissions to modify the code.

**Comment by smangancap** [comment]
The last modification to the Zoom plugin v5.2.3 indicates:

Regression: **recordingtype** column was too small for some type strings #605 (thanks @samwitzig, @acquaalta)
Introduced in v5.2.1 when fixing recording type language strings.

Could this have anything to do with it?

It appears as though the type is: **Shared screen with speaker view** as per my post above.

**Comment by smangancap** [comment]
Another update - just heard back from our vendor and he is willing to update the code to debug for us.

**Comment by smangancap** [comment]
> ```
> if (empty($recording->recording_type)) {
>     var_dump($recording);
> }
> ```

Can you please confirm which file and this entry goes? Would it be the file that cron is tripping up on or the webservice file? Sorry, I'm not a developer :)

/mod/zoom/classes/webservice.php
or
/mod/zoom/classes/task/get_meeting_recordings.php

Also, where specifically we should put it. Thank you.

**Comment by jrchamp** [comment]
In `webservice.php`, within the `get_user_recordings()` function. It may be best for that debug code to be immediately before or immediately after the related line of code: `$recordinginfo->recordingtype = $recording->recording_type;`

**Comment by smangancap** [comment]
I couldn't recreate the problem on our development instance so we just updated our production server.

Here's the output of the cron job log after inserting that code. I also upped debugging to developer temporarily and checked the server logs and nothing specific to this was logged there.

The error still exists for a few of our courses now and is still halting all other recordings from being returned back to moodle. I will continue to work on trying to find some commonality in the zoom meeting configuration for the courses that are returning the error.

Get meeting recordings from Zoom
Execute scheduled task: Get meeting recordings from Zoom (mod_zoom\task\get_meeting_recordings)
... started 10:14:23. Current memory use 16.5 MB.
Finding meeting recordings for this account...
Recording id: 57ba7966-d337-48bf-8c5a-a77b58505d64 exists...skipping
Recording id: 8a5cf846-b84d-41a8-ba1d-244feb34821f exists...skipping
Recording id: 9db2e398-f82d-4964-997a-e121d5e0a199 exists...skipping
Recording id: ca36843f-4125-4e86-b548-58bc7ab8e8af exists...skipping
Recording id: 0873ae4a-4f71-4f06-ad6f-6ffd50da41b5 exists...skipping
Recording id: d650ae42-000b-4651-8815-fb701f387f2a exists...skipping
Meeting id: 64329308632 does not exist...skipping
Meeting id: 64329308632 does not exist...skipping
object(stdClass)#648 (10) {
  ["id"]=>
  string(36) "a4cec3a2-fa86-4183-9056-2a867a8338ca"
  ["meeting_id"]=>
  string(24) "LDTuMMSPRUW8B3Qq9a4sZg=="
  ["recording_start"]=>
  string(20) "2024-09-24T01:16:17Z"
  ["recording_end"]=>
  string(20) "2024-09-24T04:21:58Z"
  ["file_type"]=>
  string(4) "CHAT"
  ["file_extension"]=>
  string(3) "TXT"
  ["file_size"]=>
  int(1156)
  ["play_url"]=>
  string(127) "https://capu.zoom.us/rec/play/Fm9AY697Dl436dhSXrNSnTLRY7WjpPj9haWM7uBdeTCimsRRo91dBxEP_C7HFZexQ4LeaFhDKVOXQQ8H.Ef4ZISFzm8vdt9hd"
  ["download_url"]=>
  string(131) "https://capu.zoom.us/rec/download/Fm9AY697Dl436dhSXrNSnTLRY7WjpPj9haWM7uBdeTCimsRRo91dBxEP_C7HFZexQ4LeaFhDKVOXQQ8H.Ef4ZISFzm8vdt9hd"
  ["status"]=>
  string(9) "completed"
}
Recording id: 22f476d1-438c-474d-af4a-aea6ef82f0ab exists...skipping
Recording id: 81b07e9b-4f50-4033-aad9-720a210fb6d2 exists...skipping
... used 3 dbqueries
... used 7.2225430011749 seconds
Scheduled task failed: Get meeting recordings from Zoom (mod_zoom\task\get_meeting_recordings),Error writing to database
Debug info:
ERROR:  null value in column "recordingtype" violates not-null constraint
DETAIL:  Failing row contains (1856, 2748, LDTuMMSPRUW8B3Qq9a4sZg==, a4cec3a2-fa86-4183-9056-2a867a8338ca, ZOOM Class link passcode is 733671, https://capu.zoom.us/rec/play/Fm9AY697Dl436dhSXrNSnTLRY7WjpPj9ha..., tr5$9!Cj, null, 1727140577, 1, 1727198063, 1727198063).
INSERT INTO mdl_zoom_meeting_recordings (zoomid,meetinguuid,zoomrecordingid,name,externalurl,passcode,recordingtype,recordingstart,showrecording,timecreated,timemodified) VALUES($1,$2,$3,$4,$5,$6,$7,$8,$9,$10,$11) RETURNING id
[array (
  'zoomid' => '2748',
  'meetinguuid' => 'LDTuMMSPRUW8B3Qq9a4sZg==',
  'zoomrecordingid' => 'a4cec3a2-fa86-4183-9056-2a867a8338ca',
  'name' => 'ZOOM Class link passcode is 733671',
  'externalurl' => 'https://capu.zoom.us/rec/play/Fm9AY697Dl436dhSXrNSnTLRY7WjpPj9haWM7uBdeTCimsRRo91dBxEP_C7HFZexQ4LeaFhDKVOXQQ8H.Ef4ZISFzm8vdt9hd',
  'passcode' => 'tr5$9!Cj',
  'recordingtype' => NULL,
  'recordingstart' => 1727140577,
  'showrecording' => '1',
  'timecreated' => 1727198063,
  'timemodified' => 1727198063,
)]
Backtrace:
* line 293 of /lib/dml/moodle_read_slave_trait.php: call to moodle_database->query_end()
* line 341 of /lib/dml/pgsql_native_moodle_database.php: call to pgsql_native_moodle_database->read_slave_query_end()
* line 1132 of /lib/dml/pgsql_native_moodle_database.php: call to pgsql_native_moodle_database->query_end()
* line 1180 of /lib/dml/pgsql_native_moodle_database.php: call to pgsql_native_moodle_database->insert_record_raw()
* line 153 of /mod/zoom/classes/task/get_meeting_recordings.php: call to pgsql_native_moodle_database->insert_record()
* line 263 of /lib/cronlib.php: call to mod_zoom\task\get_meeting_recordings->execute()
* line 167 of /admin/cli/scheduled_task.php: call to cron_run_inner_scheduled_task()

**Comment by jrchamp** [comment]
Thank you @smangancap! It seems that `recording_type` is not always being set even though the Zoom API documentation implies that it would always to be set. That seems like a bug in Zoom's API, but also something that we should have a fallback in place for.

If you are able, please apply the fix from #620. This should allow the data to at least load (and display the type as "null"). When/if the recording_type data becomes available from Zoom, the displayed string should automatically correct itself to match Zoom. I'll try to report the issue to Zoom's API team and hopefully it will stop happening.

**Comment by jrchamp** [comment]
A few days ago, we were able to release this fix as part of [v5.2.4](https://github.com/ncstate-delta/moodle-mod_zoom/releases/tag/v5.2.4). Sorry for the delay. ❤️
